loopd/daemon.go

```diff
@@ -38,72 +42,85 @@ type listenerCfg struct {
 	getLnd func(string, *lndConfig) (*lndclient.GrpcLndServices, error)
 }
 
-// daemon runs loopd in daemon mode. It will listen for grpc connections,
-// execute commands and pass back swap status information.
-func daemon(config *config, lisCfg *listenerCfg) error {
-	lnd, err := lisCfg.getLnd(config.Network, config.Lnd)
-	if err != nil {
-		return err
-	}
-	defer lnd.Close()
+// Daemon is the struct that holds one instance of the loop client daemon.
+type Daemon struct {
+	// To be used atomically. Declared first to optimize struct alignment.
+	started int32
+	stopped int32
 
-	// If no swap server is specified, use the default addresses for mainnet
-	// and testnet.
-	if config.SwapServer == "" {
-		// TODO(wilmer): Use onion service addresses when proxy is
-		// active.
-		switch config.Network {
-		case "mainnet":
-			config.SwapServer = mainnetServer
-		case "testnet":
-			config.SwapServer = testnetServer
-		default:
-			return errors.New("no swap server address specified")
-		}
+	// swapClientServer is the embedded RPC server that satisfies the client
+	// RPC interface. We embed this struct so the Daemon itself can be
+	// registered to an existing grpc.Server to run as a subserver in the
+	// same process.
+	swapClientServer
+
+	cfg         *Config
+	listenerCfg *listenerCfg
+
+	lnd           *lndclient.GrpcLndServices
+	clientCleanup func()
+
+	wg sync.WaitGroup
+
+	mainCtx       context.Context
+	mainCtxCancel func()
+
+	grpcServer    *grpc.Server
+	grpcListener  net.Listener
+	restServer    *http.Server
+	restListener  net.Listener
+	restCtxCancel func()
+}
+
+// New creates a new instance of the loop client daemon.
+func New(config *Config, lisCfg *listenerCfg) *Daemon {
+	return &Daemon{
+		cfg:         config,
+		listenerCfg: lisCfg,
 	}
+}
 
-	log.Infof("Swap server address: %v", config.SwapServer)
+// Start starts loopd in daemon mode. It will listen for grpc connections,
+// execute commands and pass back swap status information.
+func (d *Daemon) Start() error {
+	// There should be no reason to start the daemon twice. Therefore return
+	// an error if that's tried. This is mostly to guard against Start and
+	// StartAsSubserver both being called.
+	if atomic.AddInt32(&d.started, 1) != 1 {
+		return errOnlyStartOnce
+	}
 
-	// Create an instance of the loop client library.
-	swapClient, cleanup, err := getClient(config, &lnd.LndServices)
+	var err error
+	d.lnd, err = d.listenerCfg.getLnd(d.cfg.Network, d.cfg.Lnd)
 	if err != nil {
+		_ = d.Stop()
```

Comment:
There is a lot of stop calling on error. Probably everywhere. Create a wrapping function that calls this function, inspects the error and call stop if it is not nil?